### PR TITLE
Fixing up inference performance bug for diffusion-based EBT

### PR DESCRIPTION
**Pull request by travisddavies**
Fixing a bug where training performance is great but inference performance is terrible. Basically the gradient being calculated is a scalar rather than a batch of gradients. This means that during training the EBT model is learning to predict energies/gradients based on the entire batch, so during inference when the batch size is 1, the performance drops to horrible levels.

Below is a PR to fix the issue

### CODE REVIEW

**Comment by travisddavies** [comment]
Using a single gradient scalar calculated off a whole batch to update the predicted embeddings batch means that the batch samples are not independent during training. This results in a disconnect between training performance and inference performance, since the batch size will change. 

My evidence is that I introduced this code into an imitation learning model and had great training and validation performance. Euclidean distance for predicted actions was roughly 0.0005 m. The picture below represents ground truth x, y, z coordinates of a robotic arm and EBT's predicted trajectory.

sampled-action

Below is EBT's performance during inference, when batch size is 1.

inference

After changing the code to what I wrote, this is the inference performance:

new-inference

So based on this evidence I do believe that it's the fact that the gradient over a batch is being calculated to a single scalar rather than a vector of gradients.

However maybe investigate for yourself, maybe it was just in my case.

**Comment by alexiglad** [comment]
First off, that's super exciting, and the results look really cool!

Before I do some more checks, some questions for you:
Are you using batch norm or any other metrics that depend on the batch? With the transformers we trained we made sure to not use any functions that depend on batch metrics, meaning doing energy.sum() doesn't mix information across a batch. If you do use activation functions or anything that uses batch-level metrics (i.e. often contrastive losses like InfoNCE) then this will not be the case.

**Comment by travisddavies** [comment]
Thanks! It's actually very exciting findings, I'm gonna run it on a robot this week or next!

Since I'm using an EMA model during training there's only RMSNorm and QKNorm being used. Activation functions are GELU. So shouldn't be anything to do with that.

**Comment by amorehead** [comment]
@travisddavies, how does your change interact with the `is_grads_batched` argument of https://docs.pytorch.org/docs/stable/generated/torch.autograd.grad.html?

**Comment by travisddavies** [comment]
> Ok so two things:
> 
> * I believe the only line actually needed in your PR (functionally) are lines 117-118, can you verify that this is the case?
> * Can you print the tensor shape on line 118? I believe you'll be collapsing to a scalar in this case, unless your tensor are different somehow. If you are collapsing to a scalar this will make it hard to collect the energies of different samples.
> 
> Thanks once again for your help, really appreciate it!

Shape at line 118 is 32 (batch size)

After running with just the change at 117-118 it seems to have solved the issue as well. Thanks for that!

**Comment by alexiglad** [comment]
@travisddavies, what is the dimension of your data (i.e. x)? Is it B, S, D; or just B, D?

**Comment by travisddavies** [comment]
> @travisddavies, what is the dimension of your data (i.e. x)? Is it B, S, D; or just B, D?

B, S, D

**Comment by sergiosolorzano** [comment]
I would 🤩

**Comment by travisddavies** [comment]
> Would love to!!

@alexiglad where can you receive a video?

**Comment by travisddavies** [comment]
> @travisddavies I think I may know the issue. The output after `energy_preds = self.transformer(predicted_embeddings, condition)` should be B, S, 1. After squeezing it, it should be B, S; and then after taking a mean and reshaping it'll be just B. I believe the output of your energy_preds is B, S, D instead. Can you confirm if this is the case?
> 
> If it is, which architecture are you using? Does the final layer map down the embedding dim to 1?

Okay I can have a look for you today

**Comment by alexiglad** [comment]
@travisddavies feel free to email me and others who reacted to your message :)
alexigladstone@gmail.com

**Comment by sergiosolorzano** [comment]
Thank you for the heads up @travisddavies and the offering to share @travisddavies :) if you could please email it to me at sergiosolorzano@gmail.com Thank you! Sergio

**Comment by travisddavies** [comment]
> @travisddavies I think I may know the issue. The output after `energy_preds = self.transformer(predicted_embeddings, condition)` should be B, S, 1. After squeezing it, it should be B, S; and then after taking a mean and reshaping it'll be just B. I believe the output of your energy_preds is B, S, D instead. Can you confirm if this is the case?
> 
> If it is, which architecture are you using? Does the final layer map down the embedding dim to 1?

Ah yes it seems that this is the issue, since the length for my last output is 128, not 1. Which is why my model was averaging over the batch dimension. Yep makes sense
